pg_shard--1.2.sql

```diff
@@ -1,53 +1,140 @@
-	CREATE SEQUENCE shard_id_sequence MINVALUE 10000 NO CYCLE
-	CREATE SEQUENCE shard_placement_id_sequence NO CYCLE;
-
--- mark each of the above as config tables to have pg_dump preserve them
-SELECT pg_catalog.pg_extension_config_dump(
-	'pgs_distribution_metadata.shard', '');
-SELECT pg_catalog.pg_extension_config_dump(
-	'pgs_distribution_metadata.shard_placement', '');
-SELECT pg_catalog.pg_extension_config_dump(
-	'pgs_distribution_metadata.partition', '');
+-- needed in our views
+CREATE FUNCTION column_to_column_name(table_oid oid, column_var text)
+RETURNS text
+AS 'MODULE_PATHNAME'
+LANGUAGE C STABLE STRICT;
+
+DO $$
+DECLARE
+    replication_factor integer;
+    use_citus_metadata boolean := false;
+BEGIN
+    BEGIN
+        replication_factor = current_setting('shard_replication_factor')::integer;
+        use_citus_metadata = (replication_factor IS NOT NULL);
+    EXCEPTION
+        WHEN undefined_object THEN
+            use_citus_metadata = false;
+    END;
+
+    IF use_citus_metadata THEN
+        CREATE FUNCTION adapt_and_insert_shard_placement() RETURNS trigger AS $aaisp$
+        BEGIN
+            INSERT INTO pg_dist_shard_placement
+                        (shardid,
+                         shardstate,
+                         shardlength,
+                         nodename,
+                         nodeport)
+            VALUES      (NEW.shard_id,
+                         NEW.shard_state,
+                         0,
+                         NEW.node_name,
+                         NEW.node_port);
+
+          RETURN NEW;
+        END
+        $aaisp$ LANGUAGE plpgsql;
+
+        CREATE FUNCTION adapt_and_insert_partition() RETURNS trigger AS $aaip$
+        BEGIN
+            INSERT INTO pg_dist_partition
+                        (logicalrelid,
+                         partmethod,
+                         partkey)
+            VALUES      (NEW.relation_id,
+                         NEW.partition_method,
+                         column_name_to_column(NEW.relation_id, NEW.key));
+
+          RETURN NEW;
+        END
+        $aaip$ LANGUAGE plpgsql;
+
+        -- we still provide this schema to help migrations
+        CREATE SCHEMA pgs_distribution_metadata
+            CREATE VIEW shard AS
+                SELECT shardid       AS id,
+                       logicalrelid  AS relation_id,
+                       shardstorage  AS storage,
+                       shardminvalue AS min_value,
+                       shardmaxvalue AS max_value
+                FROM   pg_dist_shard
+
+            CREATE VIEW shard_placement AS
+                SELECT oid::bigint AS id,
+                       shardid     AS shard_id,
+                       shardstate  AS shard_state,
+                       nodename    AS node_name,
+                       nodeport    AS node_port
+                FROM   pg_dist_shard_placement
+
+            CREATE TRIGGER shard_placement_insert INSTEAD OF INSERT ON shard_placement
```

Comment:
Will deletes/updates error out? Or should we have triggers similar to the insert.